Add developer documentation (#86)
UCL/TDMS: doc/developers.md

* [MATLAB](https://www.mathworks.com/products/matlab.html) since we read in and write out MATLAB format files. We are actually thinking of removing this as a strict dependency ([#70](https://github.com/UCL/TDMS/issues/70)).

### Code style and other admin

We try to stick to [Google style C++](https://google.github.io/styleguide/cppguide.html) wherever practicable. <!-- And we run `clang-format` and `cppclean` linters as part of our CI. -->
There is one exception: we prefer `#pragma once` at the top of each header (because it's one line rather than three and is almost the standard anyway).

Please keep doxygen-style comments in the **headers**.
```{.cpp}
/**
 * Add two numbers together
 *
 * @param a: real number
 * @param b: real number
 * @return sum of a and b
 */
double add(double a, double b);
```
And please document the header itself with at least:
```{.cpp}
/**
 * @file filename.h
 * 
 * Describe the functions/classes in here.
 */
 #pragma once
```
You can quickly check the documentation for your changes look correct with:
```{.sh}
doxygen doc/Doxygen

> firefox html/index.html # you should be able to find and read your documentation

Reviewer: guess we should cater to the rogue firefox users out there..

```
Don't worry about doxygen for the source files (although obviously please do write helpful comments there).

### Compiling and debugging

- By default, build testing is turned off. You can turn it on with `-DBUILD_TESTING=ON`.

> - Also by default debug printout is off. Turn on with `-DCMAKE_BUILD_TYPE=Test` or manually at some specific place in the code with:

Reviewer: Beeds 
```suggestion
- Also by default debug printout is off. Turn on with `-DCMAKE_BUILD_TYPE=Debug` or manually at some specific place in the code with:
```
now given @giordano's [suggestion](https://github.com/UCL/TDMS/pull/83#discussion_r965829660)

> # Developer documentation

Reviewer 2: If this site has API docs, then won't users use it? Maybe in a future PR "Developer documentation" can become a sub-page of the whole site.

Author: Depends. Users who just want to compile this then run their MATLAB scripts, no. Although how many of them are we expecting? Unsure.

Reviewer 2: 👍 fair enough - I don't know how users interact with the code so wasn't sure if the API docs was more dev or user focused, but it sounds like dev and therefore this is fine.

Author: > Maybe in a future PR "Developer documentation" can become a sub-page of the whole site.

I had been thinking of this though. So I'm glad I'm not the only one.

Reviewer: as far as I'm aware the only user interaction with the c++ code is `tdms <input_file> <output_file>`

Welcome to the developer documentation for TDMS, the Time Domain Maxwell Solver.

> These pages contain the C++ API [doxygen](https://doxygen.nl) documentation as well as some other useful information for developers.

Reviewer 2: ```suggestion
These pages contain the C++ API documentation as well as some other useful information for developers.
```

- 🐣 If you are a new user, you probably want to start with the project's [README.md](https://github.com/UCL/TDMS/blob/main/README.md) on github. There are installation and "getting started" instructions.

> - 🐛 If you've spotted a bug or want to request a feature, we have a short [CONTRIBUTING.md](https://github.com/UCL/TDMS/blob/main/CONTRIBUTING.md). Then please [send an issue through github](https://github.com/UCL/TDMS/issues).

Reviewer 2: ```suggestion
- 🐛 If you've spotted a bug or want to request a feature, read [CONTRIBUTING.md](https://github.com/UCL/TDMS/blob/main/CONTRIBUTING.md). Then please raise an issue on the [GitHub issue tracker](https://github.com/UCL/TDMS/issues).
```

> - If you want to contribute to the development (🚀!) and have read [CONTRIBUTING.md](https://github.com/UCL/TDMS/blob/main/CONTRIBUTING.md), you're in the right place.

Reviewer 2: ```suggestion
- If you want to contribute to the development (🚀!) and have read [CONTRIBUTING.md](https://github.com/UCL/TDMS/blob/main/CONTRIBUTING.md), keep reading this page for more information.
```

## What?

> The physics and method used by TDMS is described in [this pdf document](https://github.com/UCL/TDMS/blob/gh-doc/masterdoc.pdf) (the source is in [doc/latex](https://github.com/UCL/TDMS/blob/main/doc/latex)).

Reviewer 2: ```suggestion
The physics background and numerical methods used by TDMS are described in [this pdf document](https://github.com/UCL/TDMS/blob/gh-doc/masterdoc.pdf) (the source is in [doc/latex](https://github.com/UCL/TDMS/blob/main/doc/latex)).
```

In terms of the code, `tdms` is a C++ binary that can be installed in your path.

> We build with [CMake](https://cmake.org/) to be moderately platform-independent. We run CI builds and tests in Windows, Ubuntu, and MacOS.

Reviewer 2: ```suggestion
We build with [CMake](https://cmake.org/) to be moderately platform-independent. We run continuous integration builds and tests on Windows, Ubuntu, and MacOS.
```

### Code style and other admin

We try to stick to [Google style C++](https://google.github.io/styleguide/cppguide.html) wherever practicable. <!-- And we run `clang-format` and `cppclean` linters as part of our CI. -->
There is one exception: we prefer `#pragma once` at the top of each header (because it's one line rather than three and is almost the standard anyway).

Please keep doxygen-style comments in the **headers**.
```{.cpp}
/**
 * Add two numbers together
 *
 * @param a: real number
 * @param b: real number
 * @return sum of a and b
 */
double add(double a, double b);
```
And please document the header itself with at least:
```{.cpp}
/**
 * @file filename.h
 * 
 * Describe the functions/classes in here.
 */
 #pragma once
```
You can quickly check the documentation for your changes look correct with:
```{.sh}
doxygen doc/Doxygen
firefox html/index.html # or your web browser of choice
```
You should be able to find and read what you've changed.
Don't worry about doxygen for the source files (although obviously please do write helpful comments there).

### Compiling and debugging

> 

Reviewer 2: Add a line here with instructions to compile (e.g. `mkdir build; cd build; cmake ..; make` or whatever works for TDMS)?

Author: It's on the README. So I kinda didn't want to duplicate it. But you're probably right.

Reviewer 2: Yeah, splitting bits of the same documentation over two different sources isn't ideal. Perhaps all the developer focused stuff could be stripped out of the README and into the new site, and README.md could be more user facing?
